quasar-osgi-tests/src/osgi-test/java/org/testing/osgi/BundleExclusionTest.java

```diff
@@ -0,0 +1,109 @@
+package org.testing.osgi;
+
+import co.paralleluniverse.fibers.Suspendable;
+import co.paralleluniverse.fibers.suspend.Instrumented;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.test.common.annotation.InjectBundleContext;
+import org.osgi.test.common.annotation.InjectService;
+import org.osgi.test.junit5.context.BundleContextExtension;
+import org.osgi.test.junit5.service.ServiceExtension;
+import org.testing.osgi.security.SecurityConfig;
+import org.testing.osgi.unprivileged.Unprivileged;
+
+import java.io.InputStream;
+import java.lang.reflect.Method;
+import java.util.concurrent.Callable;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+import static org.testing.osgi.security.SecurityConfig.ADMIN_PERMISSIONS;
+import static org.testing.osgi.security.SecurityConfig.ALL_PERMISSIONS;
+
+@ExtendWith({ BundleContextExtension.class, ServiceExtension.class })
+@TestInstance(PER_CLASS)
+class BundleExclusionTest {
+    private static final String CALLABLE_CLASS_NAME = "org.testing.osgi.suspendable.ExampleCallable";
+    private static final String MESSAGE = "Hello Quasar!";
+
+    @InjectBundleContext
+    BundleContext bundleContext;
+
+    @BeforeAll
+    void setup(
+        @InjectService(timeout = 1000)
+        SecurityConfig securityConfig
+    ) {
+        securityConfig.setSecurityPolicy(
+            // Any call-stack containing Unprivileged has only these permissions.
+            securityConfig.allowFor(Unprivileged.class, ADMIN_PERMISSIONS),
+            securityConfig.denyAllFor(Unprivileged.class),
+
+            // Everyone else has all permissions.
+            securityConfig.allow("*", ALL_PERMISSIONS)
+        );
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = { "IGNORE/", "VERIFY/" })
+    void testBundleNotInstrumented(String locationPrefix) throws Exception {
+        Bundle excluded = bundleContext.installBundle(withLocationPrefix(locationPrefix), getSuspendableJar());
```

Comment:
Does `bundleContext.installBundle` close the stream?

Reply:
Yes. See [BundleContext.getInputStream()](http://docs.osgi.org/javadoc/osgi.core/7.0.0/org/osgi/framework/BundleContext.html#installBundle-java.lang.String-java.io.InputStream-)
> input - The InputStream object from which this bundle will be read or null to indicate the Framework must create the input stream from the specified location identifier. **The input stream must always be closed when this method completes, even if an exception is thrown.**

Reply:
Thanks. I saw that documentation but wasn't sure if that was an instruction for us or a definition of behaviour.